README.md
## Web Server Middleware for Dart

[![CI](https://github.com/dart-lang/shelf/workflows/CI/badge.svg?branch=master)](https://github.com/dart-lang/shelf/actions?query=workflow%3ACI+branch%3Amaster)

## Introduction

**Shelf** makes it easy to create and compose **web servers** and **parts of web
servers**. How?

- Expose a small set of simple types.
- Map server logic into a simple function: a single argument for the request,
  the response is the return value.
- Trivially mix and match synchronous and asynchronous processing.
- Flexibility to return a simple string or a byte stream with the same model.

See the
[Dart HTTP server documentation](https://dart.dev/tutorials/server/httpserver)
for more information. You may also want to look at
[package:shelf_router](https://pub.dev/packages/shelf_router) and
[package:shelf_static](https://pub.dev/packages/shelf_static) as examples of
packages that build on and extend `package:shelf`.

## Example

See `example/example.dart`

```dart
import 'package:shelf/shelf.dart';
import 'package:shelf/shelf_io.dart' as shelf_io;

void main() async {
  var handler =
      const Pipeline().addMiddleware(logRequests()).addHandler(_echoRequest);

  var server = await shelf_io.serve(handler, 'localhost', 8080);

  // Enable content compression
  server.autoCompress = true;

  print('Serving at http://${server.address.host}:${server.port}');
}

Response _echoRequest(Request request) =>
    Response.ok('Request for "${request.url}"');
```

## Handlers and Middleware

A [Handler][] is any function that handles a [Request][] and returns a
[Response][]. It can either handle the request itself–for example, a static file
server that looks up the requested URI on the filesystem–or it can do some
processing and forward it to another handler–for example, a logger that prints
information about requests and responses to the command line.

[handler]: https://pub.dev/documentation/shelf/latest/shelf/Handler.html
[request]: https://pub.dev/documentation/shelf/latest/shelf/Request-class.html
[response]: https://pub.dev/documentation/shelf/latest/shelf/Response-class.html

The latter kind of handler is called "[middleware][]", since it sits in the
middle of the server stack. Middleware can be thought of as a function that
takes a handler and wraps it in another handler to provide additional
functionality. A Shelf application is usually composed of many layers of
middleware with one or more handlers at the very center; the [Pipeline][] class
makes this sort of application easy to construct.

[middleware]: https://pub.dev/documentation/shelf/latest/shelf/Middleware.html
[pipeline]: https://pub.dev/documentation/shelf/latest/shelf/Pipeline-class.html

Some middleware can also take multiple handlers and call one or more of them for
each request. For example, a routing middleware might choose which handler to
call based on the request's URI or HTTP method, while a cascading middleware
might call each one in sequence until one returns a successful response.

Middleware that routes requests between handlers should be sure to update each
request's [`handlerPath`][handlerpath] and [`url`][url]. This allows inner
handlers to know where they are in the application so they can do their own
routing correctly. This can be easily accomplished using
[`Request.change()`][change]:

[handlerpath]:
  https://pub.dev/documentation/shelf/latest/shelf/Request/handlerPath.html
[url]: https://pub.dev/documentation/shelf/latest/shelf/Request/url.html
[change]: https://pub.dev/documentation/shelf/latest/shelf/Request/change.html

```dart
// In an imaginary routing middleware...
var component = request.url.pathSegments.first;
var handler = _handlers[component];
if (handler == null) return Response.notFound(null);

// Create a new request just like this one but with whatever URL comes after
// [component] instead.
return handler(request.change(path: component));
```

## Adapters

An adapter is any code that creates [Request][] objects, passes them to a
handler, and deals with the resulting [Response][]. For the most part, adapters
forward requests from and responses to an underlying HTTP server;
[shelf_io.serve][] is this sort of adapter. An adapter might also synthesize
HTTP requests within the browser using `window.location` and `window.history`,
or it might pipe requests directly from an HTTP client to a Shelf handler.

[shelf_io.serve]: https://pub.dev/documentation/shelf/latest/shelf_io/serve.html

### API Requirements

An adapter must handle all errors from the handler, including the handler
returning a `null` response. It should print each error to the console if
possible, then act as though the handler returned a 500 response. The adapter
may include body data for the 500 response, but this body data must not include
information about the error that occurred. This ensures that unexpected errors
don't result in exposing internal information in production by default; if the
user wants to return detailed error descriptions, they should explicitly include
middleware to do so.

An adapter should ensure that asynchronous errors thrown by the handler don't
cause the application to crash, even if they aren't reported by the future
chain. Specifically, these errors shouldn't be passed to the root zone's error
handler; however, if the adapter is run within another error zone, it should
allow these errors to be passed to that zone. The following function can be used
to capture only errors that would otherwise be top-leveled:

```dart
/// Run [callback] and capture any errors that would otherwise be top-leveled.
///
/// If [this] is called in a non-root error zone, it will just run [callback]
/// and return the result. Otherwise, it will capture any errors using
/// [runZoned] and pass them to [onError].
void catchTopLevelErrors(
  void Function() callback,
  void Function(Object error, StackTrace stackTrace) onError,
) {
  if (Zone.current.inSameErrorZone(Zone.root)) {
    return runZonedGuarded(callback, onError);
  } else {
    return callback();
  }
}
```

An adapter that knows its own URL should provide an implementation of the
[`Server`][server] interface.

[server]: https://pub.dev/documentation/shelf/latest/shelf/Server-class.html

### Request Requirements

When implementing an adapter, some rules must be followed. The adapter must not
pass the `url` or `handlerPath` parameters to [Request][]; it should only pass
`requestedUri`. If it passes the `context` parameter, all keys must begin with
the adapter's package name followed by a period. If multiple headers with the
same name are received, the adapter must collapse them into a single header
separated by commas as per [RFC 2616 section 4.2][].

[rfc 2616 section 4.2]: https://www.w3.org/Protocols/rfc2616/rfc2616-sec4.html

If the underlying request uses a chunked transfer coding, the adapter must
decode the body before passing it to [Request][] and should remove the
`Transfer-Encoding` header. This ensures that message bodies are chunked if and
only if the headers declare that they are.

### Response Requirements

An adapter must not add or modify any [entity headers][] for a response.

[entity headers]: https://www.w3.org/Protocols/rfc2616/rfc2616-sec7.html#sec7.1

If _none_ of the following conditions are true, the adapter must apply [chunked
transfer coding][] to a response's body and set its Transfer-Encoding header to
`chunked`:

- The status code is less than 200, or equal to 204 or 304.
- A Content-Length header is provided.
- The Content-Type header indicates the MIME type `multipart/byteranges`.
- The Transfer-Encoding header is set to anything other than `identity`.

[chunked transfer coding]:
  https://www.w3.org/Protocols/rfc2616/rfc2616-sec3.html#sec3.6.1

Adapters may find the [`addChunkedEncoding()`][addchunkedencoding] middleware
useful for implementing this behavior, if the underlying server doesn't
implement it manually.

[addchunkedencoding]:
  https://pub.dev/documentation/shelf/latest/shelf/addChunkedEncoding.html

When responding to a HEAD request, the adapter must not emit an entity body.
Otherwise, it shouldn't modify the entity body in any way.

An adapter should include information about itself in the Server header of the
response by default. If the handler returns a response with the Server header
set, that must take precedence over the adapter's default header.

An adapter should include the Date header with the time the handler returns a
response. If the handler returns a response with the Date header set, that must
take precedence.

## Inspiration

- [Connect](https://github.com/senchalabs/connect) for NodeJS.
- [Rack](https://github.com/rack/rack) for Ruby.


CHANGELOG.md
## 1.3.0

* Add Response.badRequest() constructor
* Deprecate `ServerHandler`.
* Require Dart >= 2.16

## 1.2.0

* Added `MiddlewareExtensions` which provides `addMiddleware` and `addHandler`
  to `Middleware` instances. Provides easier access to the composition
  capabilities offered by `Pipeline`.

## 1.1.4

* Documentation improvements.

## 1.1.3

* Automatically remove `content-length` header from a `Response.notModified`.
  Restores some of the safety around malformed requests that was removed in
  `1.1.2` where we started allowing `content-length` for some responses without
  bodies.
* Documentation cleanup.

## 1.1.2

* Allow an explicit non-zero content-length header if the body is empty.
  (Needed to correctly support HEAD requests).

## 1.1.1

* Avoid wrapping response bodies that already contained `List<int>` or
  `Stream<List<int>>` in a `CastList`/`CastStream` to improve
  performance.

## 1.1.0

* Change `Request.hijack` return type from `void` to `Never`. This may cause
  analysis hints in packages using this method, but is not actually breaking
  for apps.

## 1.0.0

* Stable null safety release.

## 1.0.0-nullsafety.0

* Update to support Dart null-safety.
* Updated the `change` function on `Request` and `Response` to clear existing
  values in `context` if the provided `context` value has `null` values.

## 0.7.9

* Allow a `handlerPath` to lead up to a double `//` in a URI.

## 0.7.8

* Handle malformed URLs (400 instead of 500).

## 0.7.7

* Fix internal error in `Request` when `.change()` matches the full path.

## 0.7.6

*   Supports multiple header values in `Request` and `Response`:
    *   `headersAll` field contains all the header values
    *   `headers` field contains the same values as previously (appending values
        with `,`)
    *   `headers` parameter in the constructor and in the `.change()` method
        accepts both `String` and `List<String>` values.

## 0.7.5

* Return the correct HTTP status code for badly formatted requests.

## 0.7.4+1

* Allow `stream_channel` version 2.x

## 0.7.4

* Allow passing `shared` parameter to underlying `HttpServer.bind` calls via
  `shelf_io.serve`.
* Correctly pass `encoding` in `Response` constructors `forbidden`, `notFound`,
  and `internalServerError`.
* Update `README.md` to point to latest docs.

## 0.7.3+3

* Set max SDK version to `<3.0.0`, and adjust other dependencies.

## 0.7.3+2

* Fix constant evaluation analyzer error in `shelf_unmodifiable_map.dart`.

* Update usage of HTTP constants from the Dart SDK. Now require 2.0.0-dev.61.

## 0.7.3+1

* Updated SDK version to 2.0.0-dev.55.0.

## 0.7.3

* Fix some Dart 2 runtime errors.

## 0.7.2

* Update `createMiddleware` arguments to use `FutureOr`.

  * Note: this change is not breaking for the runtime behavior, but it might
    cause new errors during static analysis due the the type changes.

* Updated minimum Dart SDK to `1.24.0`, which added `FutureOr`.

* Improved formatting of the `logRequests` default logger.

## 0.7.1

* The `shelf_io` server now adds a `shelf.io.connection_info` field to
  `Request.context`, which provides access to the underlying
  `HttpConnectionInfo` object.

## 0.7.0

* Give a return type to the `Handler` typedef. This may cause static warnings
  where there previously were none, but all handlers should have already been
  returning a `Response` or `Future<Response>`.

* Remove `HijackCallback` and `OnHijackCallback` typedefs.

* **Breaking**: Change type of `onHijack` in the `Request` constructor to take
  an argument of `StreamChannel`.

## 0.6.8

* Add a `securityContext` parameter to `self_io.serve()`.

## 0.6.7+2

* Go back to auto-generating a `Content-Length` header when the length is known
  ahead-of-time *and* the user hasn't explicitly specified a `Transfer-Encoding:
  chunked`.

* Clarify adapter requirements around transfer codings.

* Make `shelf_io` consistent with the clarified adapter requirements. In
  particular, it removes the `Transfer-Encoding` header from chunked requests,
  and it doesn't apply additional chunking to responses that already declare
  `Transfer-Encoding: chunked`.

## 0.6.7+1

* Never auto-generate a `Content-Length` header.

## 0.6.7

* Add `Request.isEmpty` and `Response.isEmpty` getters which indicate whether a
  message has an empty body.

* Don't automatically generate `Content-Length` headers on messages where they
  may not be allowed.

* User-specified `Content-Length` headers now always take precedence over
  automatically-generated headers.

## 0.6.6

* Allow `List<int>`s to be passed as request or response bodies.

* Requests and responses now automatically set `Content-Length` headers when the
  body length is known ahead of time.

* Work around [sdk#27660][] by manually setting
  `HttpResponse.chunkedTransferEncoding` to `false` for requests known to have
  no content.

[sdk#27660]: https://github.com/dart-lang/sdk/issues/27660

## 0.6.5+3

* Improve the documentation of `logRequests()`.

## 0.6.5+2

* Support `http_parser` 3.0.0.

## 0.6.5+1

* Fix all strong-mode warnings and errors.

## 0.6.5

* `Request.hijack()` now takes a callback that accepts a single `StreamChannel`
  argument rather than separate `Stream` and `StreamSink` arguments. The old
  callback signature is still supported, but should be considered deprecated.

* The `HijackCallback` and `OnHijackCallback` typedefs are deprecated.

## 0.6.4+3

* Support `http_parser` 2.0.0.

## 0.6.4+2

* Fix a bug where the `Content-Type` header didn't interact properly with the
  `encoding` parameter for `new Request()` and `new Response()` if it wasn't
  lowercase.

## 0.6.4+1

* When the `shelf_io` adapter detects an error, print the request context as
  well as the error itself.

## 0.6.4

* Add a `Server` interface representing an adapter that knows its own URL.

* Add a `ServerHandler` class that exposes a `Server` backed by a `Handler`.

* Add an `IOServer` class that implements `Server` in terms of `dart:io`'s

lib/shelf_io.dart
// Copyright (c) 2014, the Dart project authors.  Please see the AUTHORS file
// for details. All rights reserved. Use of this source code is governed by a
// BSD-style license that can be found in the LICENSE file.

/// A Shelf adapter for handling [HttpRequest] objects from `dart:io`'s
/// [HttpServer].
///
/// One can provide an instance of [HttpServer] as the `requests` parameter in
/// [serveRequests].
///
/// This adapter supports request hijacking; see [Request.hijack].
///
/// [Request]s passed to a [Handler] will contain the [Request.context] key
/// `"shelf.io.connection_info"` containing the [HttpConnectionInfo] object from
/// the underlying [HttpRequest].
///
/// When creating [Response] instances for this adapter, you can set the
/// `"shelf.io.buffer_output"` key in [Response.context]. If `true`,
/// (the default), streamed responses will be buffered to improve performance.
/// If `false`, all chunks will be pushed over the wire as they're received.
/// See [HttpResponse.bufferOutput] for more information.
import 'dart:async';
import 'dart:io';

import 'package:collection/collection.dart';
import 'package:http_parser/http_parser.dart';
import 'package:stack_trace/stack_trace.dart';
import 'package:stream_channel/stream_channel.dart';

import 'shelf.dart';
import 'src/util.dart';

export 'src/io_server.dart' show IOServer;

/// Starts an [HttpServer] that listens on the specified [address] and
/// [port] and sends requests to [handler].
///
/// If a [securityContext] is provided an HTTPS server will be started.
///
/// See the documentation for [HttpServer.bind] and [HttpServer.bindSecure]
/// for more details on [address], [port], [backlog], and [shared].
Future<HttpServer> serve(
  Handler handler,
  address,
  int port, {
  SecurityContext? securityContext,
  int? backlog,
  bool shared = false,
}) async {
  backlog ??= 0;
  var server = await (securityContext == null
      ? HttpServer.bind(address, port, backlog: backlog, shared: shared)
      : HttpServer.bindSecure(
          address,
          port,
          securityContext,
          backlog: backlog,
          shared: shared,
        ));
  serveRequests(server, handler);
  return server;
}

/// Serve a [Stream] of [HttpRequest]s.
///
/// [HttpServer] implements [Stream<HttpRequest>] so it can be passed directly
/// to [serveRequests].
///
/// Errors thrown by [handler] while serving a request will be printed to the
/// console and cause a 500 response with no body. Errors thrown asynchronously
/// by [handler] will be printed to the console or, if there's an active error
/// zone, passed to that zone.
void serveRequests(Stream<HttpRequest> requests, Handler handler) {
  catchTopLevelErrors(() {
    requests.listen((request) => handleRequest(request, handler));
  }, (error, stackTrace) {
    _logTopLevelError('Asynchronous error\n$error', stackTrace);
  });
}

/// Uses [handler] to handle [request].
///
/// Returns a [Future] which completes when the request has been handled.
Future handleRequest(HttpRequest request, Handler handler) async {
  Request shelfRequest;
  try {
    shelfRequest = _fromHttpRequest(request);
  } on ArgumentError catch (error, stackTrace) {
    if (error.name == 'method' || error.name == 'requestedUri') {
      // TODO: use a reduced log level when using package:logging
      _logTopLevelError('Error parsing request.\n$error', stackTrace);
      final response = Response(
        400,
        body: 'Bad Request',
        headers: {HttpHeaders.contentTypeHeader: 'text/plain'},
      );
      await _writeResponse(response, request.response);
    } else {
      _logTopLevelError('Error parsing request.\n$error', stackTrace);
      final response = Response.internalServerError();
      await _writeResponse(response, request.response);
    }
    return;
  } catch (error, stackTrace) {
    _logTopLevelError('Error parsing request.\n$error', stackTrace);
    final response = Response.internalServerError();
    await _writeResponse(response, request.response);
    return;
  }

  // TODO(nweiz): abstract out hijack handling to make it easier to implement an
  // adapter.
  Response? response;
  try {
    response = await handler(shelfRequest);
  } on HijackException catch (error, stackTrace) {
    // A HijackException should bypass the response-writing logic entirely.
    if (!shelfRequest.canHijack) return;

    // If the request wasn't hijacked, we shouldn't be seeing this exception.
    response = _logError(
      shelfRequest,
      "Caught HijackException, but the request wasn't hijacked.",
      stackTrace,
    );
  } catch (error, stackTrace) {
    response = _logError(
      shelfRequest,
      'Error thrown by handler.\n$error',
      stackTrace,
    );
  }

  if ((response as dynamic) == null) {
    // Handle nulls flowing from opt-out code
    await _writeResponse(
        _logError(
            shelfRequest, 'null response from handler.', StackTrace.current),
        request.response);
    return;
  }
  if (shelfRequest.canHijack) {
    await _writeResponse(response, request.response);
    return;
  }

  var message = StringBuffer()
    ..writeln('Got a response for hijacked request '
        '${shelfRequest.method} ${shelfRequest.requestedUri}:')
    ..writeln(response.statusCode);
  response.headers.forEach((key, value) => message.writeln('$key: $value'));
  throw Exception(message.toString().trim());
}

/// Creates a new [Request] from the provided [HttpRequest].
Request _fromHttpRequest(HttpRequest request) {
  var headers = <String, List<String>>{};
  request.headers.forEach((k, v) {
    headers[k] = v;
  });

  // Remove the Transfer-Encoding header per the adapter requirements.
  headers.remove(HttpHeaders.transferEncodingHeader);

  void onHijack(void Function(StreamChannel<List<int>>) callback) {
    request.response
        .detachSocket(writeHeaders: false)
        .then((socket) => callback(StreamChannel(socket, socket)));
  }

  return Request(
    request.method,
    request.requestedUri,
    protocolVersion: request.protocolVersion,
    headers: headers,
    body: request,
    onHijack: onHijack,
    context: {'shelf.io.connection_info': request.connectionInfo!},
  );
}

Future _writeResponse(Response response, HttpResponse httpResponse) {
  if (response.context.containsKey('shelf.io.buffer_output')) {
    httpResponse.bufferOutput =
        response.context['shelf.io.buffer_output'] as bool;
  }

  httpResponse.statusCode = response.statusCode;

  // An adapter must not add or modify the `Transfer-Encoding` parameter, but
  // the Dart SDK sets it by default. Set this before we fill in
  // [response.headers] so that the user or Shelf can explicitly override it if
  // necessary.
  httpResponse.headers.chunkedTransferEncoding = false;

  response.headersAll.forEach((header, value) {
    httpResponse.headers.set(header, value);
  });

  var coding = response.headers['transfer-encoding'];
  if (coding != null && !equalsIgnoreAsciiCase(coding, 'identity')) {
    // If the response is already in a chunked encoding, de-chunk it because
    // otherwise `dart:io` will try to add another layer of chunking.
    //
    // TODO(nweiz): Do this more cleanly when sdk#27886 is fixed.
    response = response.change(
      body: chunkedCoding.decoder.bind(response.read()),
    );
    httpResponse.headers.set(HttpHeaders.transferEncodingHeader, 'chunked');
  } else if (response.statusCode >= 200 &&
      response.statusCode != 204 &&
      response.statusCode != 304 &&
      response.contentLength == null &&
      response.mimeType != 'multipart/byteranges') {
    // If the response isn't chunked yet and there's no other way to tell its
    // length, enable `dart:io`'s chunked encoding.
    httpResponse.headers.set(HttpHeaders.transferEncodingHeader, 'chunked');
  }

  if (!response.headers.containsKey(HttpHeaders.serverHeader)) {
    httpResponse.headers.set(HttpHeaders.serverHeader, 'dart:io with Shelf');
  }

  if (!response.headers.containsKey(HttpHeaders.dateHeader)) {
    httpResponse.headers.date = DateTime.now().toUtc();
  }

  return httpResponse
      .addStream(response.read())
      .then((_) => httpResponse.close());
}

// TODO(kevmoo) A developer mode is needed to include error info in response
// TODO(kevmoo) Make error output plugable. stderr, logging, etc
Response _logError(Request request, String message, StackTrace stackTrace) {
  // Add information about the request itself.
  var buffer = StringBuffer();
  buffer.write('${request.method} ${request.requestedUri.path}');
  if (request.requestedUri.query.isNotEmpty) {
    buffer.write('?${request.requestedUri.query}');
  }
  buffer.writeln();
  buffer.write(message);

  _logTopLevelError(buffer.toString(), stackTrace);
  return Response.internalServerError();
}

void _logTopLevelError(String message, StackTrace stackTrace) {
  final chain = Chain.forTrace(stackTrace)
      .foldFrames((frame) => frame.isCore || frame.package == 'shelf')
      .terse;

  stderr.writeln('ERROR - ${DateTime.now()}');
  stderr.writeln(message);
  stderr.writeln(chain);
}
